qiskit/circuit/parametertable.py

```diff
@@ -32,6 +35,8 @@ def __init__(self, *args, **kwargs):
         self._names = {x.name for x in self._table}
 
     def __getitem__(self, key):
+        if isinstance(key, int):
+            return next(islice(self._table.keys(), key, key+1))
```

Comment:
I dont fully understand this. if `self._table` is a dict without order, why accessing by numeric index.

I also would suggest this:
```suggestion
            return list(self._table.keys())[key]
```

Reply:
Since dicts are insertion ordered since python 3.5 this iterator approach to indexing is faster than generating a list and indexing that as I understand it.

Reply:
I changed the `QuantumCircuit.assign_parameters` method to use `QuantumCircuit.parameters` instead of indexing the parameter table so this is not needed anymore 🙂